Procurement/Utility/ClientLogFileWatcher.cs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        public delegate void ClientLogFileEventHandler(ClientLogFileWatcher sender, ClientLogFileEventArgs e);
+        public static event ClientLogFileEventHandler ClientLogFileChanged;
+
+        public DateTime LastDateTimeSeen
+        {
+            get;
+            protected set;
+        }
+
+        public long LastTimestampSeen
+        {
+            get;
+            protected set;
+        }
+
+        public long LastFileSizeSeen
+        {
+            get;
+            protected set;
+        }
+
+        protected System.Timers.Timer PollingTimer
+        {
+            get;
+            set;
+        }
+
+        protected Regex LocationChangedRegex
+        {
+            get;
+            set;
+        }
+
+        protected void Initialize()
+        {
+            if (!Settings.UserSettings.Keys.Contains("ClientLogFileLocation"))
+                return;
+            string fullFilePath = Settings.UserSettings["ClientLogFileLocation"];
+            if (string.IsNullOrWhiteSpace(fullFilePath))
+                return;
+
+            FileWatcher = new System.IO.FileSystemWatcher();
+            FileWatcher.Path = System.IO.Path.GetDirectoryName(fullFilePath);
+            FileWatcher.Filter = System.IO.Path.GetFileName(fullFilePath);
+            FileWatcher.NotifyFilter = NotifyFilters.LastWrite | NotifyFilters.Size | NotifyFilters.LastAccess;
+
+            FileWatcher.Changed += OnFileChanged;
+
+            PollingTimer = new System.Timers.Timer();
+            PollingTimer.Elapsed += (s, e) => { ReadClientLogFile(); };
+            PollingTimer.Interval = 30000;  // 30 seconds
+
+            // Regex to catch lines formatted like:
+            //   2019/06/21 19:16:37 245842781 aa1 [INFO Client 19844] : You have entered Sunspire Hideout.
+            LocationChangedRegex = new Regex(
+                @"(\d{4}/\d{2}/\d{2} \d{2}:\d{2}:\d{2}) (\d+) [^ .]* \[.*\] : You have entered (.*).$",
+                RegexOptions.Compiled);
+        }
+
+        internal void Start()
+        {
+            if (!Settings.UserSettings.Keys.Contains("EnableClientLogFileMonitoring"))
+                return;
+            var enabled = Convert.ToBoolean(Settings.UserSettings["EnableClientLogFileMonitoring"]);
+            if (!enabled)
+                return;
+
+            if (FileWatcher == null)
+                Initialize();
+
+            FileWatcher.EnableRaisingEvents = true;
+
+            if (!PollingTimer.Enabled)
+                PollingTimer.Start();
+        }
+
+        internal void Stop()
+        {
+            if (FileWatcher != null)
+                FileWatcher.EnableRaisingEvents = false;
+
+            PollingTimer.Stop();
+        }
+
+        protected void ReadClientLogFile()
+        {
+            lock (Instance)
+            {
+                try
+                {
+                    Stream stream = new FileStream(Settings.UserSettings["ClientLogFileLocation"], FileMode.Open,
```

Comment:
Stream needs to be in a using, this is currently not being disposed. Use nested usings:

https://stackoverflow.com/questions/1329739/nested-using-statements-in-c-sharp#1329765